### ISSUE TITLE
How to configure to match over active/passive cluster

### ISSUE THREAD

**Issue by toni-moreno** [grobian/carbon-c-relay#41]
After reviewing the documentation I don't know how to send over a cluster in fail-over mode.

consider this cluster configuration

```
cluster A
    ip_A_datacenter1
    ip_A_datacenter2

cluster B
    ip_B_datacenter1
    ip_B_datacenter2

cluster C
    ip_C_datacenter1
    ip_C_datacenter2

match "^metricstoA\.*"
    send to A

match "^metricstoB\.*"
   send to B

match "^metricstoC\.*"
    send to C

```

What I need is only send to the first host on each cluster (ip_X_datacenter1) and only if this one fails switch to the other ( ip_X_datacenter2) also should check ip_X_datacenter1 availability to come back to the main active server after the host has been recovered.

How can I do that?

**Comment by grobian** [comment]
Currently you cannot do this.  It would require a "failover" cluster.  This is not too hard, it would be a variation of the current any_of cluster.

**Comment by toni-moreno** [comment]
what files should I modify if I would like to add this new feature? 

is there any documentation about internal design ? any advice on how to do that?

**Comment by grobian** [comment]
This will definitely affect router.c (router_readconfig) for the new cluster type and various structs/enums there, and it will need some changes to router_route_intern to apply the failover logic.  Since the server targets themself "ping" to see if they are up again after failure, that should do most of the trick, and you'd want to use the secondaries feature from any_of in server.c to handle events from the server queue when it determines it went down (I think).  Tricky thing there is how to handle a queue overflow scenario which any_of does.  I think you don't want that here, so some more tweaking to that code needs to be done to make it do like you do.

Long story short, you're invited to give it a try, but I think you better let me do it for you ;)

**Comment by toni-moreno** [comment]
Hi @grobian I'll be pleased to test it if you can do it for me!!
I'm now evaluating also other tools to do that , I will install at some "production servers" in a few weeks if available.

**Comment by grobian** [comment]
Can you please give the latest head a try?  No guarantees :)

**Comment by toni-moreno** [comment]
Great !!
Hi @grobian , this is  a good news for us, we will  do some test today !!!

**Comment by grobian** [comment]
thanks

**Comment by toni-moreno** [comment]
Hi @grobian firts test are OK,   with a very basic configuration (only one failover cluster ), when first server goes down second one began to receive metrics from carbon-c-cache.

```
cluster A
    failover
        192.168.150.143:2003
        192.168.150.144:2003
    ;

match *
    send to A
    stop
    ;
```

We are now doing more test but we have some questions about carbon-c-relay behavior.

* I've seen that it don't maintain tcp  connection opened if  no metrics should be sent. Can I force maintain tcp connection open ( let us get  where is currently sending data with lsof ).

* we can not see self statistics metrics anywhere with previous configuration, why? how should I configure to send them to the cluster A ?

* where carbon-c-relay logs all the debug output ? ( what file?).

* when -s used I can see output log on stdout, and no probes seems to be done on the passive node until it begins to work on it. isn't it?


```
root@graphiteCollectd:~# /usr/local/sbin/carbon-c-relay -s -f /usr/local/etc/carbon/carbon-c-relay.conf
[2015-01-26 11:22:31] starting carbon-c-relay v0.37 (2cf849)
configuration:
    relay hostname = graphiteCollectd
    listen port = 2003
    workers = 2
    send batch size = 2500
    server queue size = 25000
    statistics submission interval = 60s
    submission = true
    routes configuration = /usr/local/etc/carbon/carbon-c-relay.conf

parsed configuration follows:
cluster A
    failover
        192.168.150.143:2003
        192.168.150.144:2003
    ;

match *
    send to A
    stop
    ;

listening on tcp4 0.0.0.0 port 2003
listening on tcp6 :: port 2003
listening on udp4 0.0.0.0 port 2003
listening on udp6 :: port 2003
listening on UNIX socket /tmp/.s.carbon-c-relay.2003
starting 2 workers
starting statistics collector

[2015-01-26 11:29:21] failed to connect() to 192.168.150.143:2003: Connection refused
[2015-01-26 11:31:01] server 192.168.150.143:2003: OK after probe
```

**Comment by grobian** [comment]
1) no you can't unless you change the code to increase the timeout
2) self metrics aren't sent because you use -s
3) there is no debug output
4) probes are done always when the node is in failure state. When the relay starts, it has no reason to contact any node until it receives data (it assumes all nodes to be "up") to be sent.

**Comment by toni-moreno** [comment]
@grobian I will maintain this issue opened for some weeks until a full test and load test done. do you agree?

**Comment by grobian** [comment]
yeah, sure, no problem

**Comment by grobian** [comment]
@toni-moreno if you're still using this, does it work to your satisfaction?

**Comment by toni-moreno** [comment]
Hi @grobian , I'm still in testing stage we have plans to build a big production environment on 2 months and carbon-c-relay will be  in the front of a group of graphites doing routing / failover tasks.

all  tests are ok now.

You can close this issue if you want  , I will do further test on the final infrastructure , and I will report you the results.

Thank you very much

**Comment by grobian** [comment]
Ok, that sounds exciting!  Looking forward to hear about how it goes.  Closing this issue then as it appears ok.
